### PR TITLE
Migrating file info store cache to the cache layer

### PR DESCRIPTION
**Pull request by jespino**
#### Summary
Migrating file info store cache to the cache layer

#### Ticket
[MM-20047](https://mattermost.atlassian.net/browse/MM-20047)

### CODE REVIEW

**Comment by jespino** [comment]
@lindy65 sorry, there is no jira ticket for this, but I'll create one and link it here.

**Comment by lindy65** [comment]
Thanks @jespino - appreciated!

**Comment by jespino** [comment]
@lindy65 Done

File: store/localcachelayer/file_info_layer_test.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+	t.Run("first call not cached, second cached and returning same data", func(t *testing.T) {
+		mockStore := getMockStore()
+		cachedStore := NewLocalCacheLayer(mockStore, nil, nil)
+
+		fileInfos, err := cachedStore.FileInfo().GetForPost("123", true, true, true)
+		require.Nil(t, err)
+		assert.Equal(t, fileInfos, []*model.FileInfo{&fakeFileInfo})
+		mockStore.FileInfo().(*mocks.FileInfoStore).AssertNumberOfCalls(t, "GetForPost", 1)
+		require.Nil(t, err)
```
Comment (rvillablanca) [inline store/localcachelayer/file_info_layer_test.go]: I don't understand why you are asserting ´err´ and ´fileInfos´again if they weren't assigned :thinking:

Comment (jespino) [reply store/localcachelayer/file_info_layer_test.go]: Good, point, is a mistake, is not needed.

File: store/localcachelayer/file_info_layer_test.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+		cachedStore := NewLocalCacheLayer(mockStore, nil, nil)
+
+		cachedStore.FileInfo().GetForPost("123", true, true, true)
+		mockStore.FileInfo().(*mocks.FileInfoStore).AssertNumberOfCalls(t, "GetForPost", 1)
+		cachedStore.FileInfo().InvalidateFileInfosForPostCache("123", true)
+		cachedStore.FileInfo().GetForPost("123", true, true, true)
+		mockStore.FileInfo().(*mocks.FileInfoStore).AssertNumberOfCalls(t, "GetForPost", 2)
+	})
+
+	t.Run("first call not cached, second cached and returning same data", func(t *testing.T) {
```
Comment (rvillablanca) [inline store/localcachelayer/file_info_layer_test.go]: This step is repeated with the first

File: store/localcachelayer/file_info_layer_test.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+		mockStore := getMockStore()
+		cachedStore := NewLocalCacheLayer(mockStore, nil, nil)
+
+		cachedStore.FileInfo().GetForPost("123", true, true, true)
+		mockStore.FileInfo().(*mocks.FileInfoStore).AssertNumberOfCalls(t, "GetForPost", 1)
+		cachedStore.FileInfo().GetForPost("123", true, true, false)
+		mockStore.FileInfo().(*mocks.FileInfoStore).AssertNumberOfCalls(t, "GetForPost", 2)
+	})
+
+	t.Run("first call not cached, invalidate, and then not cached again", func(t *testing.T) {
```
Comment (rvillablanca) [inline store/localcachelayer/file_info_layer_test.go]: This step is repeated too

**Comment by rvillablanca** [comment]
I think some steps in the tests were repeated by mistake

**Comment by jespino** [comment]
@rvillablanca Thanks for the feedback, now is fixed :)

**Comment by lieut-data** [comment]
@marianunez, did we still want to meet with @jespino to get an overview of the "behind-the-scenes" changes here and so provide a better code review?

**Comment by marianunez** [comment]
> @marianunez, did we still want to meet with @jespino to get an overview of the "behind-the-scenes" changes here and so provide a better code review?

@lieut-data Yes, I discussed with @jespino and I believe its a matter of finding some time to synch on this.

File: store/sqlstore/file_info_store.go [hunk 4]
```diff
@@ -111,35 +99,10 @@ func (fs SqlFileInfoStore) GetByPath(path string) (*model.FileInfo, *model.AppEr
 	return info, nil
 }
 
-func (fs SqlFileInfoStore) InvalidateFileInfosForPostCache(postId string) {
-	fileInfoCache.Remove(postId)
-	if fs.metrics != nil {
-		fs.metrics.IncrementMemCacheInvalidationCounter("File Info Cache - Remove by PostId")
-	}
+func (fs SqlFileInfoStore) InvalidateFileInfosForPostCache(postId string, deleted bool) {
```
Comment (marianunez) [inline store/sqlstore/file_info_store.go]: Do we need to keep this here? It seems that now it should be calling the one in `LocalCacheFileInfoStore`

File: store/timer_layer.go [hunk 5]
```diff
@@ -3722,6 +3736,22 @@ func (s *TimerLayerPluginStore) SaveOrUpdate(keyVal *model.PluginKeyValue) (*mod
 	return resultVar0, resultVar1
 }
 
+func (s *TimerLayerPluginStore) SetWithOptions(pluginId string, key string, value interface{}, options model.PluginKVSetOptions) (bool, *model.AppError) {
```
Comment (marianunez) [inline store/timer_layer.go]: Is this related to the File Info store cache?

File: store/timer_layer.go [hunk 6]
```diff
@@ -2938,6 +2936,22 @@ func (s *TimerLayerGroupStore) GetMemberUsersPage(groupID string, page int, perP
 	return resultVar0, resultVar1
 }
 
+func (s *TimerLayerGroupStore) PermanentDeleteMembersByUser(userId string) *model.AppError {
```
Comment (marianunez) [inline store/timer_layer.go]: Is this related to the File Info store cache?

File: app/post.go [hunk 7]
```diff
@@ -1116,7 +1116,8 @@ func (a *App) GetFileInfosForPostWithMigration(postId string) ([]*model.FileInfo
 		post := result.Data.(*model.Post)
 
 		if len(post.Filenames) > 0 {
-			a.Srv.Store.FileInfo().InvalidateFileInfosForPostCache(postId)
+			a.Srv.Store.FileInfo().InvalidateFileInfosForPostCache(postId, false)
+			a.Srv.Store.FileInfo().InvalidateFileInfosForPostCache(postId, true)
```
Comment (marianunez) [inline app/post.go]: Why do we call this twice? Should we always just invalidate for deleted as well?

Also, shouldn't we remove from the FileInfo cache when `DeleteForPost` is called?

File: store/localcachelayer/layer.go [hunk 8]
```diff
@@ -20,6 +20,9 @@ const (
 	SCHEME_CACHE_SIZE = 20000
 	SCHEME_CACHE_SEC  = 30 * 60
 
+	FILE_INFO_CACHE_SIZE = 25000
+	FILE_INFO_CACHE_SEC  = 30 * 60
+
```
[the remark this answers is not in the document]
Comment (jespino) [reply store/localcachelayer/layer.go]: They are independent and can be different, so I prefer to have there some degree of duplication instead of making hard to tune this things.

File: store/sqlstore/file_info_store.go (hunk 4 above)
Comment (jespino) [reply store/sqlstore/file_info_store.go]: Yes, because is part of the Store interface, so we need to implement it (in this case with an empty implementation).

File: store/timer_layer.go (hunk 6 above)
Comment (jespino) [reply store/timer_layer.go]: Not really, but is generated because there is this new store API endpoint, is not necesarily something that should be here, but is harmless, anyway, I don't mind to removed it from here and add it again in another PR.

File: store/timer_layer.go (hunk 5 above)
Comment (jespino) [reply store/timer_layer.go]: Same as before.

File: app/post.go (hunk 7 above)
Comment (jespino) [reply app/post.go]: I guess so, but wasn't there before, so I'm only migrating the existing implementation in the new cache layer, but I'm ok adding that to other places, like `PermanentDelete`, `PermanentDeleteBatch`, `PermanentDeleteByUser` or `AttachToPost`. Anyway, we can add a different task for that.

**Comment by jespino** [comment]
@lieut-data @marianunez I going to add other reviewers to the PR to speed up the process, anyway I going to explain the new cache layer when is finished and everything working as expected.

**Comment by agnivade** [comment]
Is this still a WIP ?

**Comment by jespino** [comment]
@agnivade no, it isn't.

**Comment by agnivade** [comment]
I will take some time to review this.

File: store/localcachelayer/file_info_layer.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+type LocalCacheFileInfoStore struct {
+	store.FileInfoStore
+	rootStore *LocalCacheStore
+}
+
+func (s *LocalCacheFileInfoStore) handleClusterInvalidateFileInfo(msg *model.ClusterMessage) {
+	if msg.Data == CLEAR_CACHE_MESSAGE_DATA {
+		s.rootStore.fileInfoCache.Purge()
+	} else {
```
Comment (agnivade) [inline store/localcachelayer/file_info_layer.go]: Remove else block and outdent. Return early from the previous if condition.

File: store/localcachelayer/layer.go [hunk 10]
```diff
@@ -227,6 +239,9 @@ func (s *LocalCacheStore) doClearCacheCluster(cache *utils.Cache) {
 
 func (s *LocalCacheStore) Invalidate() {
 	s.doClearCacheCluster(s.reactionCache)
+	s.doClearCacheCluster(s.schemeCache)
```
Comment (agnivade) [inline store/localcachelayer/layer.go]: It is not very clear to me why we are clearing schemeCache and roleCache too. Was it just a bug, and you are fixing it here ?

Comment (jespino) [reply store/localcachelayer/layer.go]: Yes

**Comment by jespino** [comment]
@marianunez we have already two dev reviews here, I think I addressed all your concerns.
